akka-stream-tests/src/test/scala/akka/stream/scaladsl/AggregateWithTimeBoundarySpec.scala

```diff
@@ -0,0 +1,233 @@
+
+    val result = Source.fromPublisher(p)
+      .via(
+        new AggregateWithTimeBoundary[Int, ListBuffer[Int], Seq[Int]](
+          allocate = ListBuffer.empty,
+          aggregate = (buffer, i) => {
+            buffer.addOne(i)
+            false
+          },
+          harvest = seq => seq.toSeq,
+          maxGap = Some(maxGap), // elements with longer gap will put put to next aggregator
+          maxDuration = None,
+          currentTimeMs = schedulerTimeMs,
+          interval = 1.milli
+        )
+      )
+      .runWith(Sink.collection)
+
+    p.sendNext(1)
+    timePasses(maxGap/2) // less than maxGap should not cause emit
+    p.sendNext(2)
+    timePasses(maxGap)
+
+    p.sendNext(3)
+    timePasses(maxGap/2) // less than maxGap should not cause emit
+    p.sendNext(4)
+    timePasses(maxGap)
+
+    p.sendNext(5)
+    timePasses(maxGap/2) // less than maxGap should not cause emit
+    p.sendNext(6)
+    timePasses(maxGap/2) // less than maxGap should not cause emit and it does not accumulate
+    p.sendNext(7)
+    p.sendComplete()
+
+    Await.result(result, 10.seconds) should be (Seq(Seq(1, 2), Seq(3, 4), Seq(5, 6, 7)))
+
+  }
+
+  "split aggregator by total duration" in {
+    val maxDuration = 400.seconds
+
+    val p = TestPublisher.probe[Int]()
+
+    val result = Source.fromPublisher(p)
+      .via(
+        new AggregateWithTimeBoundary[Int, ListBuffer[Int], Seq[Int]](
+          allocate = ListBuffer.empty,
+          aggregate = (buffer, i) => {
+            buffer.addOne(i)
+            false
+          },
+          harvest = seq => seq.toSeq,
+          maxGap = None,
+          maxDuration = Some(maxDuration), // elements with longer gap will put put to next aggregator
+          currentTimeMs = schedulerTimeMs,
+          interval = 1.milli
+        )
+      )
+      .runWith(Sink.collection)
+
+    p.sendNext(1)
+    timePasses(maxDuration/4)
+    p.sendNext(2)
+    timePasses(maxDuration/4)
+    p.sendNext(3)
+    timePasses(maxDuration/4)
+    p.sendNext(4)
+    timePasses(maxDuration/4) // maxDuration will accumulate and reach threshold here
+
+    p.sendNext(5)
+    p.sendNext(6)
+    p.sendNext(7)
+    p.sendComplete()
+
+    Await.result(result, 10.seconds) should be (Seq(Seq(1, 2, 3, 4), Seq(5, 6, 7)))
+
+  }
+
+  "down stream back pressure should not miss data on completion with pull on start" in {
+
+    val maxGap = 1.second
+    val upstream = TestPublisher.probe[Int]()
+    val downstream = TestSubscriber.probe[Seq[Int]]()
+
+    Source.fromPublisher(upstream).via(
+      new AggregateWithTimeBoundary[Int, ListBuffer[Int], Seq[Int]](
+        allocate = ListBuffer.empty,
+        aggregate = (buffer, i) => {
+          buffer.addOne(i)
+          false
+        },
+        harvest = seq => seq.toSeq,
+        maxGap = Some(maxGap),
+        maxDuration = None,
+        currentTimeMs = schedulerTimeMs,
+        interval = 1.milli
+      )
+    ).via(Buffer(1, OverflowStrategy.backpressure))
```

Comment:
@johanandren great suggestion for composing with the Buffer operator.
I used this test to verify that composing the current minimum implementation (without eager pull and internal buffer) with the Buffer operator (with OverflowStrategy.backpressure) is equivalent to the previous more complex implementation with eager pull and internal buffer. The eager pull behavior led me to implementing internal buffer in the first place. So indeed eager pull is not necessary either as a whole package. Now this all makes better sense to me. Please let me know what do you think.